mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -61,6 +61,55 @@ pub struct ClientConfig {
 
     /// The region
     region: Option<String>,
+
+    /// List of network interfaces to be used by the client.
+    ///
+    /// The client will determine the logic for balancing requests over the network interfaces
+    /// according to its own logic.
+    network_interface_names: Option<NetworkInterfaceNames>,
+}
+
+/// This struct bundles together the list of owned strings for the network interfaces, and the
+/// cursors pointing to them.
+///
+/// This is necessary because the CRT expects a pointer to the list of cursors,
+/// and the lifetime of the cursors must meet or exceed the lifetime of the [ClientConfig].
+/// By wrapping the array of cursors in this struct, we know the cursors will be valid
+/// and we can ensure the memory will be freed when dropped at the same time as [ClientConfig].
+#[derive(Debug)]
+struct NetworkInterfaceNames {
+    /// The list of network interface names.
+    ///
+    /// We use a boxed array of strings, as we have no need for a mutable list like [Vec].
+    _names: Box<[String]>,
+
+    /// List of owned cursors. This will be pointed at by [ClientConfig]'s inner [aws_s3_client_config].
+    cursors: Box<[aws_byte_cursor]>,
+}
+
+impl NetworkInterfaceNames {
+    /// Create a new [NetworkInterfaceNamesInner].
+    pub fn new(names: Vec<String>) -> Self {
+        // Turn this into a read-only representation of the vector of strings.
+        let names = names.into_boxed_slice();
+
+        let cursors = names
+            .iter()
+            .map(|name| {
+                // SAFETY: The names are stored alongside the cursors in NetworkInterfaceNamesInner.
+                //         The lifetime of NetworkInterfaceNamesInner will always meet the lifetime of the cursors.
```

Comment:
The other thing that matters is that the `String`s will never be mutated, since that could reallocate the backing memory for the string and make these cursors dangle. But as above, I think you can just do this cursor array as a temporary at construction time and then it doesn't matter.